rfcs/90-support-initialization-via-config-string-and-map.md

```diff
@@ -0,0 +1,115 @@
+---
+author: xxchan <xxchan22f@gmail.com>
+status: draft
+updated_at: 2021-06-09
+deprecates:
+  - 13-remove-config-string.md
+---
+
+# GSP-90: Support Initialization Via Config String and Map
+
+## Background
+
+We had config strings like `s3://bucket_name?access_key=xxxx&secret_key=xxxx` before. This is introduced in [GSP-3] but deprecated by [GSP-13].
+
+There are two reasons why config strings are removed, as mentioned in [GSP-13]:
+
+1. Usability: If an application uses [go-storage]'s config string, then it either exposes the format of config string to end user directly (additional burden for end users) or it has to manipulate its config into our format (additional burden for application developers). And if the application wants to combine its other config into a config string, this is also inconvenient.
+2. Implementation: We didn't find a good way to parse pairs from strings easily and safely.
+
+So we let the users of [go-storage] construct pairs directly using `WithXxx()`. 
+
+But passing string config is indeed more convenient, if not much. And actually parsing string into pairs is not that difficult. We can use a registry mechanism like [GSP-48]'s implementation.
+
+## Proposal
+
+So I propose to support service init from config string and map:
+
+We add the following APIs:
+```go
+func NewServicerFromMap(ty string, m map[string]string) (types.Servicer, error) {}
+func NewStoragerFromMap(ty string, m map[string]string) (types.Storager, error) {}
+
+func NewServicerFromString(config string) (types.Servicer, error) {}
+func NewStoragerFromString(config string) (types.Storager, error) {}
+```
+
+### Format
+
+The format of the config string is:
+
+TODO
+
+#### Parseable Value Types
+
+TODO
+
+### Implementation
+
+`New*FromString` will first parse config string into `(ty string, m map[string]string)`, and then call `New*FromMap(ty, m)`.
+
+`New*FromMap` will first parse `(ty string, m map[string]string)` into `ps []Pairs`, and then call `New*(ty, ps)`.
+To support this, we have to know that a name is a pair (global or service), and its type, so we implement:
+
+#### Pair Registry
+
+We register the types of global pairs and service pairs.
+
+From [go-storage] side:
+
+```go
+type PairMap map[string]string
+func RegisterGlobalPairMap(m PairMap) {}
```

Comment:
After the service type prefix trimmed, maybe we don't need to `RegisterGlobalPairMap` and `RegisterServicePairMap` separately anymore?

Reply:
What if service A has pair `(P, T1)`, but service B has pair `(P, T2)`, and `T1 != T2`?

If you mean all the pairs will share the same namespace after prefix trimmed, I think this may be dangerous.

Reply:
Oh, this can be checked in `init` here

Reply:
I don't think they are in the same namespace:

- If `P` is a global pair, `T1 == T2` is ensured.
- If `P` is a service pair, user must use them in `A.WihtP` and `B.WithP`, and I think it's OK to allow `T1 != T2`?

Reply:
If they are not in the same namespace, we will still need `RegisterServicePairMap`

Reply:
Well, we do need `RegisterServicePairMap`, but we can register global pairs together instead of another public API `RegisterGlobalPairMap`.

Reply:
Oh, you mean global pairs are included in any service pair map

Reply:
Yes, global pairs are shared across all services and they can be included in any service pair map. (We can check them while doing generation)